lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -0,0 +1,257 @@
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  void visitBeginAccess(BeginAccessInst *beginAccess, AccessedStorage storage,
+                        AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+  bool domByKeyPath(BeginAccessInst *dominatedInstr);
+  bool analyzeDomSubTree(SILBasicBlock *block,
+                         AccessedStorageInfo &visitedDomAccessesToStorageInfo);
+
+  bool analyze();
+  void optimize();
+
+private:
+  SILFunction &func;
+  DominanceInfo *domInfo;
+  // domPairs is a vector of (Dominating begin_access, Dominated begin_access)
+  // both accesses are dynamic, and the Dominated one has no nested conflict
+  // we can turn the dominated to static during the optimize() phase
+  DomPairSet domPairs;
+  // All stdlib entry points that we've found in this function
+  KeyPathEntryPointsSet keypathEntries;
+};
+} // namespace
+
+bool DominatedAccessRemoval::domByKeyPath(BeginAccessInst *dominatedInstr) {
+  for (SILInstruction *keyPathEntry : keypathEntries) {
+    if (domInfo->properlyDominates(keyPathEntry, dominatedInstr)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool DominatedAccessRemoval::visitInstruction(
+    SILInstruction *instr,
+    AccessedStorageInfo &visitedDomAccessesToStorageInfo) {
+  if (auto *BAI = dyn_cast<BeginAccessInst>(instr)) {
+    if (BAI->getEnforcement() != SILAccessEnforcement::Dynamic) {
+      return true;
+    }
+    AccessedStorage storage = findAccessedStorageNonNested(BAI->getSource());
+    if (!storage) {
+      return true;
+    }
+    visitBeginAccess(BAI, storage, visitedDomAccessesToStorageInfo);
+  } else if (auto fullApply = FullApplySite::isa(instr)) {
+    SILFunction *callee = fullApply.getReferencedFunction();
+    if (!callee)
+      return true;
+    if (!callee->hasSemanticsAttr("_keyPathEntryPoint"))
+      return true;
+    // we can't eliminate dominated checks even when we can prove that
+    // the dominated scope has no internal nested conflicts.
+    keypathEntries.insert(fullApply.getInstruction());
+  } else if (auto *BUAI = dyn_cast<BeginUnpairedAccessInst>(instr)) {
+    // We have an Implementation that handles this in the analyzer
+    // and optimizer in a previous commit, However,
+    // In order to simplify the code, at least in the initial version,
+    // We decided to just bail on these extremely rare cases
+    return false;
+  }
+  return true;
+}
+
+void DominatedAccessRemoval::visitBeginAccess(
+    BeginAccessInst *beginAccess, AccessedStorage storage,
+    AccessedStorageInfo &visitedDomAccessesToStorageInfo) {
+  if (!storage.isUniquelyIdentifiedOrClass()) {
+    // No cannot do anything about this location -
+    // bail on trying to turn it to static
+    return;
+  }
+
+  // checks if the current storage has never been seen before
+  auto predNewStorageLoc = [&](AccessedStoragePair it) {
+    auto currStorage = it.second;
+    return !currStorage.isDistinctFrom(storage);
```

Comment:
I think is is the wrong check. AFAIK, !isDistinct is not the inverse of locations being equal. And the equal-check is needed to be able to optimize the access.